### ISSUE TITLE
[Mac] Performance Regression v4.0.0 vs v3.0.0

### ISSUE THREAD

**Issue by mmastika** [transmission/transmission#4849]
### What is the issue?

Hi there,

Am not certain of the cause, however, I noticed the performance on download differs greatly in 4.0.0 (280ace12f8) vs 3.00 (f4489c982e).

Downloading the same torrent in the older version utilises the entire bandwidth that I allow it (no limit), whilst the newer version struggles to find and connect to peers. Am happy to provide more information however I'm not certain where to start to allow further investigation.  I'm able to complete the same torrent in less than 5 minutes whilst the new version struggles to complete 10% in the previous day. I'm reverting back to v3.0.0 for the time being.

Love the software and would love to contribute more to keep it a great software.

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.0 (280ace12f8) vs 3.00 (f4489c982e)

**Comment by complexkinch** [comment]
I agree. On my M1 Macbook Pro this new version has been "pinwheeling" and the downloads are a lot slower than normal. I had to come to Github and create a new account just to research and comment on this issue. I have been waiting for an update that fixes it. I might have to downgrade to 3.0.0.

**Comment by GEORGEGIM** [comment]
Hi,
I do agree that, as it looks, version 4.0.0 appears to be problematic. No problems with version 3.0.0 (I' m running High Sierra 10.13.6).
In addition, I detected that "resume all" button in the toolbar actually pauses everything. At the same time the "pause all" button cannot be selected.
It seems better to reinstall version 3.0.0.

**Comment by SupremeAntBee** [comment]
I am agreeing with those who have posted, so far, as to the poor performance of Version 4.0.0 vs 3.0.0.  It is so much slower and the functions are not as usable as the previous version.  I have changed nothing as far as the app's settings on my end, and nothing has been changed on my Mac's setup either.
Thank you!

**Comment by kippjohnson** [comment]
I am having the same issue. About 200 transfers going and the interface pinwheels after about every click. This is on a New Mac Mini M2 (2023) with reasonably fast external disk connected via USB-C. I am getting read/writes >200 mb/s so I don't think this is the issue, but I am not sure how else to better diagnose the cause of this issue. I just updated from V3 to V4. 

For devs: essentially any action in the GUI (scrolling, clicking file, etc.) will cause the app to pinwheel, sometimes for a second or two, sometimes for >30 seconds. The action does seem to finally complete and is worse when there are things that are actively downloading or uploading, it seems. I am checking htop and activity monitor when it's pinwheeling but it seems to only be using a few % of CPU on a single thread, so I think it is waiting on something but not sure what the issue is. If you can suggest any better ways to diagnose this please let me know.

**Comment by Coeur** [comment]
Charles said a "pretty significant fix" was done with #4826 which is scheduled for Transmission 4.0.1. Maybe that will help.

**Comment by Coeur** [comment]
> In addition, I detected that "resume all" button in the toolbar actually pauses everything. At the same time the "pause all" button cannot be selected.

Hum, @GEORGEGIM, we don't deal with multiple issues in a same GitHub issue. Can someone open a separate issue specifically about that?

**Comment by kippjohnson** [comment]
> If there is anyway you can suggest for me to help better diagnose/characterize the issue please just let me know. Only happened with upgrade to V4

**Comment by atstockland** [comment]
I am also experience significant slow downs.  I have only 2 downloads in progress (both have healthy seeds).  Before upgrading from 3 to 4, I was downloading these files in a handful of minutes.  After upgrading to v4.0---I've been watching these downloads struggle for 3 days.

pause and resume seems to kick start things again---both only lasts for a short period.  And, overall download speeds seem to be significantly limited.

**Comment by Coeur** [comment]
Can I encourage some of you to try a nightly build?

**Comment by ckerr** [comment]
Agreed, people in this ticket please try out a nightly build. https://github.com/transmission/transmission/pull/4826 may  help with this issue.

**Comment by kippjohnson** [comment]
Trying today's nightly build, I get "Transmission would like to access files on a removable drive" popup hundreds of times before I gave up clicking "OK". I have previously allowed full disk access to Transmission in settings, not sure why this is happening now. Apologies if I'm missing something obvious. External drive is not encrypted nor with any special permissions. It's APFS.

**Comment by samdoran** [comment]
> I get "Transmission would like to access files on a removable drive" popup hundreds of times before I gave up clicking "OK".

That's a side effect of using the nightly build which is not notarized.

**Comment by kippjohnson** [comment]
> > I get "Transmission would like to access files on a removable drive" popup hundreds of times before I gave up clicking "OK".
> 
> That's a side effect of using the nightly build which is not notarized.

Anyway to get around that for testing?

**Comment by samdoran** [comment]
Transmission 4 is also much slower and less responsive for me. Files do download, but I get the spinning wait cursor quite often and transfer speeds are much slower than Transmission 3.

I tested with the latest nightly (4.0.0-dev (dcd7501211)) and it didn't offer much improvement.

Transmission 3.0.0 (f4489c982e) on the same system works as expected.

```
ProductName:		macOS
ProductVersion:		13.2
BuildVersion:		22D49

Model Name: MacBook Air
Model Identifier: MacBookAir10,1
Model Number: Z128000DLLL/A
Chip: Apple M1
Total Number of Cores: 8 (4 performance and 4 efficiency)
Memory: 16 GB
```

I'm happy to provide additional information if it will aid in troubleshooting.

**Comment by samdoran** [comment]
> Anyway to get around that for testing?

@kippjohnson Saving to an internal volume rather than an external one is the only thing I know of. macOS is (recently) pretty noisy with non notarized apps.

**Comment by complexkinch** [comment]
I downgraded to 3.0.0 and problems went away. Let me know when it’s fixed!

**Comment by atstockland** [comment]
Nightly builds are so error prone during install.  It'd be amazing if the app allowed to update from nighties.  I'm stuck on this:

/usr/local/include/QtCore/qfileinfo.h:132:22: error: 'path' is unavailable: introduced in macOS 10.15

I think I'll downgrade as well and wait for 4.0.1.

Love this app!  Thanks for all you do to maintain it!

**Comment by Coeur** [comment]
> Trying today's nightly build, I get "Transmission would like to access files on a removable drive" popup hundreds of times before I gave up clicking "OK".

Nightlies aren't signed and may cause this kind of issues. You can build the project yourself from source code using Xcode, and that should address signing issues (and popups) I believe.

**Comment by ckerr** [comment]
@nevack any chance you could do some profiling on what's using CPU cycles in the macOS client? I suspect you've got more experience in xcode than I do, so I'm not sure how useful my profiling session was:

I tried loading up about 10 torrents to download in a batch to try and reproduce the beachballing issue and didn't see it.

According to the xcode profiler, libtransmission is alright -- all of its time was spent either encrypting / decrypting peer messages or testing piece checksums.

The only piece of code from `macosx/` that showed up was `[TorrentTableView drawRow]` which was a pretty high 20%, and a couple of minor NSString additions (stringForRatio at 2.4% and percentString at 2.0%).

However the system calls were quite high so I wonder if there is excess drawing going on; I'm not sure how to look for that in the mm code.

**Comment by ckerr** [comment]
@nevack any chance you could do some profiling on what's using CPU cycles in the macOS client? I suspect you've got more experience in xcode than I do, so I'm not sure how useful my profiling session was:

I tried loading up about 10 torrents to download in a batch to try and generate some workload.

- I didn't see any beachballing on my 6yo Macbook.

- According to the xcode profiler, libtransmission is doing OK -- all of its time was spent either encrypting / decrypting peer messages or testing piece checksums. That's as it should be.

- The only piece of code from `macosx/` that had a significant % was `[TorrentTableView drawRow]` at 20%.

- Some of the NSString additions were higher than they should be but probably not blockers, e.g. stringForRatio at 2.4% and percentString at 2.0%.

- The system drawing calls were quite high though, so I wonder if unnecessary redraws are getting triggered. I'm not sure how to look for that in the profiler and/or mm code.

**Comment by Pentaphon** [comment]
> * I didn't see any beachballing on my 6yo Macbook.

I notice that M1 and M2 Mac users are the ones reporting the beachballing/pinwheel so you probably wouldn't see it on your older Intel Mac.

**Comment by zozulinskyi** [comment]
Hello there. I didn't have a problem with "pinwheel" but I have some problems with downloading.

In the previous version (I mean v3.0.0) I have more than 30 peers and downloading form all of them. [screen](https://snipboard.io/N8THPb.jpg)
But, in the version v4.0.0 I have only 2 peers at the same torrent file at the same time. [screen](https://snipboard.io/GMfILT.jpg)

About my system: 
 - Apple M1 Pro
 - MacOS 13.2 (22D49)

**Comment by SupremeAntBee** [comment]
> I am having the same issue. About 200 transfers going and the interface pinwheels after about every click. This is on a New Mac Mini M2 (2023) with reasonably fast external disk connected via USB-C. I am getting read/writes >200 mb/s so I don't think this is the issue, but I am not sure how else to better diagnose the cause of this issue. I just updated from V3 to V4.
> 
> For devs: essentially any action in the GUI (scrolling, clicking file, etc.) will cause the app to pinwheel, sometimes for a second or two, sometimes for >30 seconds. This primarily happens when something is being downloaded or uploaded; I just completed a download and its now scrolling without any issue.
> 
> I am checking htop and activity monitor when it's pinwheeling but it seems to only be using a few % of CPU on a single thread, so I think it is waiting on something but not sure what the issue is. If you can suggest any better ways to diagnose this please let me know.
> 
> I'm on Ventura 13.2 with Transmission 4.0.0

Sorry to hear about your and everyone’s issues with v4.0.0.  I decided to revert back to v3.0.0, and have zero issues with having done so.  The application is running as it had been before I had updated.  I’m glad about my decision to downgrade.  I hope the issues are resolved soon.  It’s a great app!

**Comment by kippjohnson** [comment]
> > Trying today's nightly build, I get "Transmission would like to access files on a removable drive" popup hundreds of times before I gave up clicking "OK".
> 
> Nightlies aren't signed and may cause this kind of issues. You can build the project yourself from source code using Xcode, and that should address signing issues (and popups) I believe.

Hi, just built from the most recent Github commit (f551b4adbff0d59557d61867d0b6518c50f5a73f) in Xcode. Fixes the "Transmission would like to access files on a removable drive" popup issue but does not rectify the pinwheeling when adding new torrents. Let me know if any further info I can add. Not an expert at Xcode's profiling tools but happy to try with some guidance.

**Comment by Coeur** [comment]
Well, the profiling tool is Instruments (right-click on the Xcode icon, Open Developer Tool, Instruments), the Template could be "Activity Monitor", then you select the Transmission app, and you do an Instruments recording starting just before the pinwheeling, and lasting for the duration of it (10-20 sec max, no need for more). Then save the recording and share the file here.

**Comment by GEORGEGIM** [comment]
Hello everybody,
Summing up my findings since I downloaded v. 4.0.0:
1. Extremely slow downloading (4 torrents remain uncompleted in the last 4 days)
2. Local data verification is required for every new-added torrent. The same applies also on several torrents already downloaded (v. 3.0.0).
3. Resume downloading can be done using the individual side button only (not from the toolbar)
4. No spinning pinwheel so far (Intel iMac)
V. 3.0.0 runs smooth and nice. None of the above issues appear.

**Comment by pallorc** [comment]
I'm also seeing the slow, laggy, pinwheeling UI in 4.0. This is on an Intel machine, MacOS 13.2. The UI seems to get less and less responsive as bandwidth use increases. Over about 10 MB/s the UI is almost unusably slow. The number of concurrent downloads doesn't seem to matter. It's more about the total bandwidth. When the download finishes, UI responsiveness returns. For the record, Transmission 3.0 was super responsive, no matter how fast it was downloading or uploading.

**Comment by diliedevs** [comment]
Same slow download issues here with v4.0.0. Also on MacBook Pro M1, macOS Ventura. Reverted back to 3.0, no problems anymore. Hopefully gets fixed soon.

**Comment by Free5ky** [comment]
Same slow download issues here with v4.0.0. 
Local data verification is required after downgrading to 3.0 
Mac Mini intel MacOS Monterey 12.4

**Comment by lukecrownet** [comment]
I'm on macOS Ventura, an Intel iMac... I haven't run into any UI slowdowns with Transmission 4.0 -- but I have started hitting massive slowdowns in download speeds... Primarily, in my case, this seems to be from most peers getting dropped. e.g., I start a new download, a dozen or more peers will connect from the swarm, Transmission will indicate it'll take about 4 minutes to download, and then peers will start dropping like flies... I'm lucky now when Transmission 4.0 is able to hold on to 1-3 peers at a time. Some downloads drop to zero peers (even though there are plenty available) and won't reconnect unless I restart Transmission -- then the cycle repeats itself. 

It's entirely consistent -- some torrents that previously would've taken only minutes to download are now taking hours or even days... But then others, every once in a while, I might still get in, say, 20 minutes instead of 3 minutes... But even in those cases, it seems to be dependent on finding 1-2 really fast peers instead of being able to pull from the full swarm.

**Comment by Yohmi** [comment]
I'm using an intel Mac. Downloading small torrents (like a few GBs) works perfectly. Today I tried to snatch an 80GB file, and unlike v3, v4.0 isn't behaving properly. When I click on the icon to display the app, it takes a few seconds before displaying the app, I get the beach ball and the activity monitor says the app isn't responding. A few seconds later it's fine, but if I click anywhere in the app, it goes back to beach ball. Memory seems totally normal (around 350MB), processor is around 60% and I think that's unusually high. I haven't been using the nightly and can't afford to break anything right now, I just wanted to confirm the malfunction on my side, while it's been working very well with smaller torrents until now (and v3 didn't have this issue either).

**Comment by ckerr** [comment]
Small update: I think I've found the cause of the beachballing. It's a little tricky but I will either ameliorate or fix it for `4.0.1` depending on how intrusive the patch is.

When `4.0.1` comes out I will probably close this ticket and ask that people file new tickets for any problems still present in `4.0.1`. This page has been great as a clearinghouse for info & I definitely appreciate the feedback. Just from a triage standpoint 4849 isn't something that can be marked as "fixed" anymore; there are so many different topics going on here (e.g. beachballing, no peers, few peers, slow DL).

**Comment by laford2** [comment]
Nightly 4.0.0-dev (3ae8107ca9) works fine. Release version 4.0 does not.  Confirming previously reported bug(s) and suggested fix.  Thank you, dev, and looking forward to 4.0.1  ;-)
(Mac OS 13.1 MBP Intel)

**Comment by sebkb** [comment]
> I downgraded to 3.0.0 and problems went away. Let me know when it’s fixed!

Same here.

**Comment by zedrdave** [comment]
Just to confirm that upgrading from 4.0 release to the nightly, instantly made my uncapped download speed go from ~1-2MB/s to ~10-20MB/s…

There is very clearly an issue with 4.0 on the Mac (and fortunately, the recent PRs have fixed it).

**Comment by wirezen** [comment]
> Small update: I think I've found the cause of the beachballing. It's a little tricky but I will either ameliorate or fix it for `4.0.1` depending on how intrusive the patch is.
> 
> When `4.0.1` comes out I will probably close this ticket and ask that people file new tickets for any problems still present in `4.0.1`. This page has been great as a clearinghouse for info & I definitely appreciate the feedback. Just from a triage standpoint 4849 isn't something that can be marked as "fixed" anymore; there are so many different topics going on here (e.g. beachballing, no peers, few peers, slow DL).

Thanks guys.
Is there an ETA for the 4.0.1 release?

**Comment by ckerr** [comment]
> Is there an ETA for the 4.0.1 release?

Wednesday

**Comment by ckerr** [comment]
For the people who were encountering beachballing and who are willing to test, please give a nightly build a spin from https://transmissionbt.com/ and LMK if it's better, worse, or no change than `4.0.0`

**Comment by Coding4Coffee** [comment]
MacOS users, can you tell me if you enabled "Randomize port on launch"? This can lead to slower speeds in certain cases. Reading some of the comments, I don't suspect this to be the issue, but satisfying my curiosity. 😸 

See: #4942

**Comment by mmastika** [comment]
@Coding4Coffee , 

Definitely not when I was on 4.0.0 before reverting back to 3.0.0.

The current settings on 3.0.0:

<img width="544" alt="Screen Shot 2023-02-19 at 1 17 43 pm" src="https://user-images.githubusercontent.com/248116/219908593-f75ed9bc-1e2f-4d6d-90c7-8e88e30da84c.png">

**Comment by Coding4Coffee** [comment]
@mmastika Can you try the nightly for 4.0 with uTP off and report back if any improvement.

**Comment by mmastika** [comment]
Hi @Coding4Coffee,

I tested the nightly build https://build.transmissionbt.com/job/trunk-mac/8844/. With uTP off, I'm happy to report that it performs better than the released 4.0.0. In the limited test that I did (with the same popular torrent) it seems to utilise the available download bandwidth much better than 3.0.0 albeit with a noticeable lag in ramping up time to max utilisation.

I was a bit curious and I did test with uTP ON with the same test parameters. Whilst the nightly build performs better than  the released 4.0.0, it wasn't able to perform as good as when the uTP is OFF. Performance was about 1/3 - 1/2.

For the purpose of this ticket however, I think it can be closed.

There are other things that I noticed, how should I report them? 

Cheers

**Comment by Coeur** [comment]
> For the purpose of this ticket however, I think it can be closed. If you could point me to the pull request/commit that fixed it, I may learn a thing or two.

Beachballing fix attempt: #4936
Downloads fix attempt: #4897

> There are other things that I noticed, how should I report them?

A separate issue for each thing that is likely going to require a separate pull request to fix. Thanks.

**Comment by pallorc** [comment]
> For the people who were encountering beachballing and who are willing to test, please give a nightly build a spin from https://transmissionbt.com/ and LMK if it's better, worse, or no change than `4.0.0`

@ckerr I tried the 3ae8107ca9 build yesterday or the day before and it lost the progress of most of the existing torrents, plus it added a bunch of new ones. The 'new' ones might have been magnets without metadata that were lost in the v3 -> v4 upgrade. Anyway the loss of so much progress state spooked me, so I quickly reverted to 4.0.0 (and the restored the app state in Application Support from a backup). All is back to normal. Didn't get a chance to test the beachballing though. Might have to test in a clean virtual machine or on another computer. But, I wanted to run that by you so you'd know about it.

**Comment by bdkjones** [comment]
## Data Point:
I replaced the production release of 4.0.0 with nightly build 4.0.0-dev (c5b1df0daa) and my issue is now resolved. I never saw a beachball or UI hang running on an M2 Max 16" MacBook Pro, but I *DID* see my download speeds drop to 1993 levels across all torrents—they were estimating more than a YEAR to complete. 

With the nightly build, they are now blazingly fast again.

**Comment by Scabbrox** [comment]
Same issue here on intel Mac.  No beach ball but speeds are much slower and larger torrents lose data and drop speed constantly taking far longer to download.

Downgrading to 3.0.0 works fine so that's what I have done for now until a release build for 4.0.0 is confirmed fixed.

**Comment by ckerr** [comment]
As mentioned [upthread](https://github.com/transmission/transmission/issues/4849#issuecomment-1433563187), several of the issues described in this ticket have been fixed by [4.0.1](https://github.com/transmission/transmission/releases/tag/4.0.1) which has now been released.

Please upgrade your systems and report new issues with as much useful detail as possible if problems persist. Thank you!

**Comment by AngeloMascaro** [comment]
I'm using v. 4.0.5 and I still have problem with the speed. It goes up and down and it takes hours to download something, while before it was matter of 3-4 minutes.
![Screen Shot 2024-01-07 at 18 44 43](https://github.com/transmission/transmission/assets/34680444/df4bfd34-79fc-41ee-893f-aa278ea269ad)

**Comment by Coeur** [comment]
@AngeloMascaro could you try a nightly build and compare? https://build.transmissionbt.com/job/trunk-mac/
And the "new" discussion thread about it is now #5261. Please move to that one.
